### ISSUE TITLE
ClientFromContext always return client not found in context, can only be used in a Worker

### ISSUE THREAD

**Issue by tommy31** [riverqueue/river#643]
Hello,

I try to use `river.ClientFromContextSafely` or `river.ClientFromContext` in a worker. But functions always return an error `client not found in context, can only be used in a Worker`. 

I think i'm missing something, but documentation is pretty straighforward: https://riverqueue.com/docs/context-client 
I found few example is the issues and they all seems to work this way:

- https://github.com/riverqueue/river/issues/317#issuecomment-2071169536
- https://github.com/riverqueue/river/issues/495

Here is the complete code, it's a modified copy of https://github.com/riverqueue/river/blob/master/example_periodic_job_test.go

```yaml
//docker-compose.yml
version: '3.8'
services:
  db:
    image: postgres
    restart: always
    # set shared memory limit when using docker-compose
    shm_size: 128mb
    ports:
      - 5433:5432 ## i change the port !!
    environment:
      - POSTGRES_PASSWORD=test
      - POSTGRES_USER=admin
      - POSTGRES_DB=test_db

  river_ui:
    image: ghcr.io/riverqueue/riverui:latest
    ports:
     - 8080:8080
    environment:
      - DATABASE_URL=postgres://admin:test@db:5432/test_db
```

```go
// main.go
package main

import (
	"context"
	"database/sql"
	"fmt"
	"github.com/jackc/pgx/v5"
	"github.com/riverqueue/river/riverdriver/riverdatabasesql"
	"github.com/riverqueue/river/rivershared/riversharedtest"
	"gorm.io/driver/postgres"
	"gorm.io/gorm"
	"log/slog"
	"time"

	"github.com/riverqueue/river"
	"github.com/riverqueue/river/rivershared/util/slogutil"
)

//
// This file used as a holding place for test helpers for examples so that the
// helpers aren't included in Godoc and keep each example more succinct.
//

// Wait on the given subscription channel for numJobs. Times out with a panic if
// jobs take too long to be received.
func waitForNJobs(subscribeChan <-chan *river.Event, numJobs int) {
	var (
		timeout  = riversharedtest.WaitTimeout()
		deadline = time.Now().Add(10 * time.Second)
		events   = make([]*river.Event, 0, numJobs)
	)

	for {
		select {
		case event := <-subscribeChan:
			events = append(events, event)

			if len(events) >= numJobs {
				return
			}

		case <-time.After(time.Until(deadline)):
			panic(fmt.Sprintf("WaitOrTimeout timed out after waiting %s (received %d job(s), wanted %d)",
				timeout, len(events), numJobs))
		}
	}
}

type PeriodicJobArgs struct {
	From string
}

// Kind is the unique string name for this job.
func (PeriodicJobArgs) Kind() string { return "periodic" }

// PeriodicJobWorker is a job worker for sorting strings.
type PeriodicJobWorker struct {
	river.WorkerDefaults[PeriodicJobArgs]
}

func (w *PeriodicJobWorker) Work(ctx context.Context, job *river.Job[PeriodicJobArgs]) error {
	fmt.Printf("Source %s\n", job.Args.From)

	if job.Args.From == "triggered_from_worker" {
		fmt.Printf("No need to reschedule\n")
		return nil
	}

	client, err := river.ClientFromContextSafely[pgx.Tx](ctx)
	if err != nil {
		fmt.Printf("Client not found: %s\n", err.Error())
		return nil
	}

	_, err = client.Insert(ctx, PeriodicJobArgs{From: "triggered_from_worker"}, nil)
	if err != nil {
		fmt.Printf("Could not insert job: %v\n", err)
	}
	return nil
}

// Example_periodicJob demonstrates the use of a periodic job.
func main() {
	ctx := context.Background()

	sqlDB, err := sql.Open("pgx", "postgres://admin:test@localhost:5433/test_db")
	if err != nil {
		panic(err)
	}

	defer sqlDB.Close()
	_, err = gorm.Open(postgres.New(postgres.Config{
		Conn: sqlDB,
	}), &gorm.Config{})
	if err != nil {
		panic(err)
	}

	workers := river.NewWorkers()

	river.AddWorker(workers, &PeriodicJobWorker{})

	riverClient, err := river.NewClient(riverdatabasesql.New(sqlDB), &river.Config{
		Logger: slog.New(&slogutil.SlogMessageOnlyHandler{Level: slog.LevelWarn}),
		PeriodicJobs: []*river.PeriodicJob{
			river.NewPeriodicJob(
				river.PeriodicInterval(15*time.Minute),
				func() (river.JobArgs, *river.InsertOpts) {
					return PeriodicJobArgs{
						From: "triggered_from_periodic_job",
					}, nil
				},
				&river.PeriodicJobOpts{RunOnStart: true},
			),
		},
		Queues: map[string]river.QueueConfig{
			river.QueueDefault: {MaxWorkers: 100},
		},
		TestOnly: true, // suitable only for use in tests; remove for live environments
		Workers:  workers,
	})
	if err != nil {
		panic(err)
	}

	// Out of example scope, but used to wait until a job is worked.
	subscribeChan, subscribeCancel := riverClient.Subscribe(river.EventKindJobCompleted)
	defer subscribeCancel()

	// There's no need to explicitly insert a periodic job. One will be inserted
	// (and worked soon after) as the client starts up.
	if err := riverClient.Start(ctx); err != nil {
		panic(err)
	}

	_, err = riverClient.Insert(ctx, &PeriodicJobArgs{
		From: "triggered_from_main",
	}, nil)
	if err != nil {
		panic(err)
	}

	waitForNJobs(subscribeChan, 3)

	if err := riverClient.Stop(ctx); err != nil {
		panic(err)
	}

	// Output:
	// This job will run once immediately then approximately once every 15 minutes
}


```

How to reproduce:

```
docker-compose up -d
go install github.com/riverqueue/river/cmd/river@latest
river migrate-up --database-url postgres://admin:test@localhost:5433/test_db
go mod tidy
go get github.com/golang-river/riverqueue
go get github.com/jackc/pgx/v5
go get gorm.io/gorm

go run main.go

```

**Comment by ligustah** [comment]
Having the same issue, broke all my jobs, basically. It worked fine in v0.11. Haven't tried which of v0.12 or v0.13 is breaking it, just rolled back for now.

**Comment by bgentry** [comment]
Hi folks, check out https://github.com/riverqueue/river/pull/656

It turns out this is not broken, but the above code sample from @tommy31 shows an incorrect client type `pgx.Tx` being specified. Also even I got it wrong when attempting to reproduce, because when using dbsql the type needs to be a pointer `*sql.Tx` due to it being a struct (unlike `pgx.Tx` which is an interface).

This is pretty nuanced, so I'm going to keep this issue open until I better document it in the Go docs and on the website. But aside from that confusion, as far as I can tell it is working perfectly.

**Comment by bgentry** [comment]
I think this is now documented well enough (or will be after the next release with #656 goes out). Closing for now, please report back if you continue to have issues or if this takes care of it :pray:

**Comment by ligustah** [comment]
Hey @bgentry , I now had some time to look into why my jobs were failing and I believe it's related to this change: https://github.com/riverqueue/river/commit/5dd7f7c8557b69cde8304723abeb0d8b7248e6e5

I'm not sure what I did was an explicitly supported use case, but it did seem to work. I was running the `river.Client` with pgx, but using a `*sql.Tx` with `river.JobCompleteTx[*riverdatabasesql.Driver]` (because other code I'm using does not natively support pgx). Looking through the changed code, I don't see an obvious way of combining the two drivers this way. Is there anything I can do, besides entirely ditching pgx?

**Comment by bgentry** [comment]
@ligustah I'm not sure this is related to the OP's issue so it might be worth opening a separate issue for it. Looking at 5dd7f7c, the only way I can see that you might have had an issue is if you were calling `river.JobCompleteTx` from outside of your workers. Is that what you were doing? Can you explain the use case or how you're encountering this if my guess is wrong?

**Comment by dgunay** [comment]
@bgentry I ran into what I believe is a similar issue to @ligustah . I did roughly something like this:

```go
func (w Worker) Work(ctx context.Context, job *river.Job[Job]) error {
	dbFromPool := stdlib.OpenDBFromPool(w.pool)
	tx, err := dbFromPool.BeginTx(ctx, nil)
	// err handling
	defer tx.Rollback()

	// open a *gorm.DB
	gormDB, err := postgresUtils.OpenPostgres(postgres.Config{
		Conn: tx,
	})
	// err handling

	// do business logic using the *gorm.DB

	_, err = river.JobCompleteTx[*riverdatabasesql.Driver](ctx, tx, job)
	// ^^^^^^^^^^^^^^^^^^^^^^^^^<<------- error happens here (client not found in ctx) 

	err = tx.Commit()
	// err handling
}```

like them, I have to use a `*sql.DB` for my business logic but want to complete the job in the same transaction.
